### PR TITLE
Add documentation for the OmegaConf grammar

### PR DESCRIPTION
**Pull request by odelalleau**
Some "design choices" I made, happy to discuss them here:
* Not dumping the whole set of parser rules as is done in [Hydra's documentation](https://hydra.cc/docs/advanced/override_grammar/basic): this is because it would be pretty much unreadable with the current documentation formatting (in particular due to the limited width). Instead, I chose to focus on a few relevant excerpts, and link to the file if someone wants to see the whole thing
* Making a few simplifications (not mentioning all details), in order to (hopefully) keep it digestible

I'm requesting reviews for diverse people to get feedback both from those familiar vs unfamiliar with it.

For ease of reviewing I suggest you build it locally:
```
cd docs && make html && open build/html/index.html
```

### CODE REVIEW

**Review by Jasha10 — COMMENTED** [review]
Very nice!

File: docs/source/grammar.rst
Comment (odelalleau) [reply docs/source/grammar.rst:67]: I can see how this may look a bit confusing. The underlying reason is that the leading dots notation for relative interpolations is "special" (those dots aren't the same as the dots used to separate a parent node from its child).

For instance `${.}` denotes the parent node (this isn't allowed anymore because it can lead to cycles in the config, but it used to work at some point). So if you think as a leading dot as meaning "parent", `${.foo}` means `${parent.foo}` and `${.[foo]}` would mean `${parent[foo]}`.

I'll let @omry chime in if he thinks we should change the current behavior. Personally I would keep it, because I like that relative interpolations are consistently using leading dots (if `${[key]}` became a relative interpolation we would lose this consistency).

Comment (Jasha10) [reply docs/source/grammar.rst:67]: This explanation makes sense to me.
Also, I think that if we make `${[key]}` invalid then there will be no easy way to access the root config with brackets notation.

Comment (Jasha10) [reply docs/source/grammar.rst:67]: This explanation makes sense to me. The "parental" dots at the beginning of the interpolation have a different meaning from the "getattr" dots that appear between keys.

File: docs/source/grammar.rst
Comment (odelalleau) [reply docs/source/grammar.rst]: Good point, this is a consequence of some over-simplifications I made. I tried to make it clearer in bb77fc1 => let me know if that's better.

I changed "string interpolation" into "interpolation string" because "string interpolation" actually means something very specific: the concatenation of at least one interpolation with either some regular string or another interpolation (which always results in a string result, e.g., `https://${host}:${port}`). But the ``text`` parser rule is also used to parse single interpolations like ``${foo}`` (that aren't string interpolations).

Comment (Jasha10) [reply docs/source/grammar.rst]: Looks good!

**Review by Jasha10 — APPROVED** [review]
Sorry I guess I forgot to give a formal approval on this one.

File: docs/source/grammar.rst
Comment (omry) [inline docs/source/grammar.rst:174]: Not sure I understand what you mean here.
In the above example you use an rstring instead of escaping it.
escaping the `\` is a Python requirement in string literals. Maybe provide an second example without an rstring calling out that this is because `\` has to be escaped in Python strings?

File: docs/source/grammar.rst
Comment (omry) [inline docs/source/grammar.rst:203]: Why do you have to escape the colon?

**Review by omry — APPROVED** [review]
Overall looks good. See some inline comments/questions.

File: docs/source/grammar.rst
Comment (odelalleau) [reply docs/source/grammar.rst]: Does e71ac8d make it clearer?

File: docs/source/grammar.rst
Comment (odelalleau) [reply docs/source/grammar.rst:209]: Ah, good question, I forgot that escaping the colon is only required in dictionary keys for obvious reasons (but escaping is optional in other unquoted strings, i.e., both `\:` and `:` lead to the same result).

I'm tempted to not go into this detail for the documentation, because it's not a big deal if people escape a colon that doesn't require it, and in most situations I expect users to just use `:` without second thoughts and it will usually work.

Let me know though if you believe this should be fully documented here.

**Comment by Jasha10** [comment]
I believe that the CI should pass if changes from PR #800 are incorporated into this branch.

**Comment by odelalleau** [comment]
> I believe that the CI should pass if changes from PR #800 are incorporated into this branch.

Thanks, I'll rebase before merging to get it green.

**Review by pixelb — APPROVED** [review]
Very nice. I had questions at the start of the change, which were all answered by the end.

I added a comment about always requiring to escape a backslash to provide forward compat,
but as for documenting the current logic, this is excellent.

thanks!

File: docs/source/grammar.rst
Comment (odelalleau) [reply docs/source/grammar.rst:189]: Yeah, that's definitely a valid point. It's a trade-off and Omry and I had some discussions about this in the past (too many discussions related to escaping to dig it up though)

There are two main reasons for the current implementation:
1. Any additional escaping of `\` can quickly make things quite unreadable when combined with Python string escaping (outside of raw strings), shell escaping, or nested interpolations. Limiting the amount of escaping to the strict minimum helps keep things "sane".
2. This keeps interpolation strings close (in behavior) to static strings, making it easier to introduce interpolations into existing strings. For instance if you had `cfg = OmegaConf.create({"msg", "Please type 'start' to get started"})` you can change this message to be dynamic with `cfg = OmegaConf.create({"msg", "Please type '${start_command}' to get started"})` without having to escape the quotes (and if somehow you had the `\'` sequence in your static string, the backslash won't suddenly disappear when you add an interpolation).

=> unless @omry wants to revisit this choice, I'd rather keep it that way :)

File: docs/source/grammar.rst
Comment (odelalleau) [reply docs/source/grammar.rst]: > Does [e71ac8d](https://github.com/omry/omegaconf/commit/e71ac8db4efb48913565fa198b99296083274d83) make it clearer?

@omry just checking if that commit looks good to you?

File: docs/source/grammar.rst
Comment (pixelb) [inline docs/source/grammar.rst:185]: This is good. Nits:

- I would avoid the last sentence as it adds noise without added understanding
- I would change "make examples more readable" to "make code more readable" to avoid any confusion that this isn't generally useful and might just be related to how the example docs are marked up or something

Comment (odelalleau) [reply docs/source/grammar.rst]: Thanks, done in 29cd492

File: docs/source/grammar.rst
Comment (odelalleau) [reply docs/source/grammar.rst]: > > Does [e71ac8d](https://github.com/omry/omegaconf/commit/e71ac8db4efb48913565fa198b99296083274d83) make it clearer?
> 
> @omry just checking if that commit looks good to you?

Actually check this diff which combines the two latest (related) commits: https://github.com/omry/omegaconf/pull/792/files/21f0217c51b562b335eb98af116690fb576bcc6e..29cd49235bfd4c88a733dd0416fb7f5e52176d5f

Comment (pixelb) [reply docs/source/grammar.rst]: This adjustment is much clearer to me. Thanks

**Pull request merged**
